### PR TITLE
Simplify CMake by using targets and add support CI build

### PR DESCRIPTION
**Pull request by uilianries**
Hello!

We are trying to package Arbiter in Conan, but we found some vendorized dependencies in this project, which totally fine.

I'm opening this PR to give the option of using all dependencies as an external component, so users are free to consume from the system.

Plus, I added a Github Workflow file, so you can check the build and test results. The build will use Conan to install all dependencies, but it could be any other package manager, including from the system. I found that the Windows build is broken due a convert error in the cpp code.

You can find the CI running on my fork: https://github.com/uilianries/arbiter/actions/workflows/ci.yml. Once this PR gets merged, it will run for any PR or Push action in your repository too. 

Using CMake targets (version +3) give a better scope control, where you can decide what should be exposed of not in terms of defines, features, flags, dependencies, etc ... But propagating properties to OBJECT targets is a bit frustrating when using separated targets (would need to copy for every CMakeLists, or keep everything as Global), so I tried to simplify to a single CMakeLists.txt. This new CMake file should have the same behavior as behavior, plus, with extra options to decide if using vendorized dependencies or not.

I hope you could consider this change.

### CODE REVIEW

File: cmake/modules/FindCurl.cmake
Comment (uilianries) [inline cmake/modules/FindCurl.cmake:1]: CMake provides official support to curl already: https://cmake.org/cmake/help/v3.13/module/FindCURL.html

File: arbiter/third/json/nlohmannjson.hpp
Comment (uilianries) [inline arbiter/third/json/nlohmannjson.hpp:1]: This is the exactly arbiter/third/json/json.hpp, only renamed now.

File: arbiter/third/json/json.hpp
Comment (uilianries) [inline arbiter/third/json/json.hpp:1]: Github diff is a bit hard, but the file content now is:

```
#ifndef ARBITER_IS_AMALGAMATION

#ifndef ARBITER_EXTERNAL_JSON
#include <arbiter/third/json/nlohmannjson.hpp>
#endif

#endif

#ifdef ARBITER_EXTERNAL_JSON
#include <nlohmann/json.hpp>
#endif

```

It's not deleted, only doing decision over the header to be included.

**Comment by hobu** [comment]
> We are trying to package Arbiter in Conan

Why? The only two significant libraries we know that use it are Entwine and PDAL, and they both vendor it. Arbiter has never had a proper release schedule, packaging setup, or versioning, and it was never intended to be a standalone library.

Appreciate the modernization of the CMake, but the total rearrangement and definition changes are disorienting.

**Comment by uilianries** [comment]
@hobu PDAL is also available in Conan. 

> Appreciate the modernization of the CMake, but the total rearrangement and definition changes are disorienting.

What about separated PRs instead? One for the CI, then another only adding options for external dependencies, no CMake targets stuff involved?

**Comment by hobu** [comment]
> @hobu PDAL is also available in Conan.

If you start distributing Arbiter as a standalone library and linking PDAL to it and then linking Entwine to it, stuff is likely to break. 

> What about separated PRs instead? One for the CI, then another only adding options for external dependencies, no CMake targets stuff involved?

That would be better, but the main point is we don't want to be responsible for maintaining a standalone, system-installed version of Arbiter anywhere. This library is about sharing code between Entwine and PDAL, not providing yet-another-remote-file-system abstraction layer for the general computing public.

**Comment by valgur** [comment]
@hobu

> That would be better, but the main point is we don't want to be responsible for maintaining a standalone, system-installed version of Arbiter anywhere.

You can continue using Arbiter under that assumption in PDAL and elsewhere regardless of whether Arbiter is usable as a standalone library.

> This library is about sharing code between Entwine and PDAL, not providing yet-another-remote-file-system abstraction layer for the general computing public.

You should definitely make that clear in the readme then, please. It looked very useful to me as a standalone library regardless of similar projects being available. Just for the fact that the AWS and GCS SDK-s are massive and can take tens of minutes to build, if nothing else.

A versioning scheme and a changelog are always welcome but they are not a strict requirement for packaging and use of the library.

**Comment by valgur** [comment]
Also
> If you start distributing Arbiter as a standalone library and linking PDAL to it and then linking Entwine to it, stuff is likely to break.

IMO it's the opposite case. Arbiter is a publicly visible dependency in PDAL and the arbiter version in PDAL and Entwine is not guaranteed to match. For Conan and other package managers, having Arbiter as a standalone package allows a consistent version to be enforced across libraries. This is the main motivation for unvendoring of Arbiter, LibXML2, nlohmann_json, etc. in the first place.

**Comment by hobu** [comment]
> IMO it's the opposite case. Arbiter is a publicly visible dependency in PDAL and the arbiter version in PDAL and Entwine is not guaranteed to match.

Can you please point to an actual issue where this is a real problem? We know that very often the vendored versions of Arbiter between Entwine and PDAL *don't* match. They're buried under the own namespaces and private so it doesn't matter.

We can address the leakage in the one public header you described in https://github.com/PDAL/PDAL/issues/4331 but given that no one uses that driver, it must not have been much of an issue.

I'm sure you can tell from our ticket traffic we are not at all enthusiastic about having the maintenance load of a bunch of fixes, ports, and config changes to unvendor the code that PDAL has vendored. They are vendored for very good reasons. 

* Eigen: at the time their API wasn't stable. This is one candidate where we could maybe unvendor it
* utfcpp: we didn't want another core external dependency. Happy to discuss this one being externally used if seen during configure.
* nanoflann: later versions are [much slower](https://github.com/microsoft/vcpkg/pull/32093#issuecomment-1602764780), the KDTree code we're using hasn't changed in years, and it is inconsistently packaged as publicly available headers when compiling (and often not available)
* kazhdan: he doesn't accept or merge fixes and it isn't in any package systems as a package. https://github.com/mkazhdan/PoissonRecon
* arbiter: it is pointless for us to accept the maintenance tail of a full library (versioning releases, release notes, etc) for some code that is shared between two projects.
* laz-perf: header-only library that we infrequently release and manage. Users and packagers complained about LASzip being effectively a hard dependency. We removed it by vendoring laz-perf.
* lepcc: good luck getting Esri to manage that library as a properly released software package. It's a minor miracle when the accept pull requests to fix actual bugs
* schema-validator: meh
* nlohmann: do you want to debug infinitely nested template compilation error messages with unknown library versions? Me neither.



*

**Pull request closed without merging**